service/history/workflow/cache/cache.go

```diff
@@ -185,8 +217,11 @@ func (c *CacheImpl) getOrCreateWorkflowExecutionInternal(
 	if !cacheHit {
 		handler.Counter(metrics.CacheMissCounter.GetMetricName()).Record(1)
 		// Let's create the workflow execution workflowCtx
-		workflowCtx = workflow.NewContext(c.shard, key, c.logger)
-		elem, err := c.PutIfNotExist(key, workflowCtx)
+		newCtx := workflow.NewContext(c.shardCtx, key, c.logger)
+		if err := c.tryRestore(ctx, c.shardCtx, newCtx); err != nil {
```

Comment:
This would be only when we fault the workflow.Context into cache, not every time we lock it. The reason I was ok with the load here is that I didn't seem like there would be many (any?) cases where we loaded the context object into cache but then _didn't_ use MutableState in some way. Does that happen?

Reply:
So the thing about lazy load is that it makes the Workflow.UpdateRegistry function pretty ugly. If we may need to restore the registry from mutable state at that point, we'll need to take a context.Context as a parameter (not _too_ bad) but we'll also need to return an error in case either the mutable state load fails or the incomplete update events can't be loaded. It's weird to have a function that should just be an accessor that needs to take a context and return an error. 

Only way around that I see would be a second implementation of update.Registry that returned an error from every call to indicate a broken state.

Reply:
I've remove update restoration from this PR pursuant to @MichaelSnowden's request above that the PR be limited in scope. I will follow up with the same thing later. FWIW, it looks like restoring the UpdateRegistry might be best done in the ConsistencyChecker which is also where we load MutableState for a given workflow.Context.

Reply:
> It's weird to have a function that should just be an accessor that needs to take a context and return an error.

I don't feel strongly about it actually, especially when comparing to other approaches I can think of.
re. consistency checker, I am not sure that's the only place workflow.Context is created.

(I guess my original concern for adding the logic inside workflow cache is that it changes the rest of the system's view on this component & how the error from it should be handled. Previously it's only for in memory locking, so timeout simply means workflow is busy, and not persistence timeout. But I guess that doesn't matter any more. :) )